feat(css): build taildwind after msbuild build phase

Js-Brecht opened the pull request:
## Description

Incorporates a couple of changes to the `.csproj` to allow for running `tailwindcss` with the `--watch` property.  Namely:

* It adds a new `<Target>` directive that only runs when the `DotNetWatch` is true
* Uses a context aware (read: operating system) `<Target>` that locates the `tailwindcss` binary

---

Additional information::

Original work was actually in the `daemon-dotnet` repo.  You can see that work [here](https://github.com/daemontechtools/daemon-dotnet/compare/main...feature/build-tasks-tailwindcss).  It's worth noting that the approach that I was following there, implementing custom tasks into MSBuild, is much more scalable/flexible, but is entirely too much for this work task.

The branch is obviously a WIP, and there's maybe another day's worth of work in testing and wiring, but we can delete it at will.  However, if any more functionality is needed in the tailwindcss department, I would suggest going that route.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

Js-Brecht committed:
set up <Exec> for watch process

Js-Brecht commented:
@JoshElias I will need to fix the extra mess in this PR in the morning.  I probably need to rebase off of `develop`.  I seem to recall seeing the history diverge a little at one point, but I may be mistaken.  Regardless, I am bushed, and I need sleep.

I also have another PR to open up tomorrow with docs for utilizing `git flow`, as well as some automated setup for some git hooks that should help enforce the `git flow` process (though we may still need to work out the PRs).

Js-Brecht commented:
nevermind, the PR was pointing at `main`... durp 🤦🏼‍♂️... after we just got done talking about that.

JoshElias commented on SmartOrder.csproj:
```diff
@@ -51,7 +48,17 @@
 
   <!-- Integrate TailwindCSS -->
-
+  <PropertyGroup Condition="$(CI) != 'true'">
+    <TailwindCssPath>/usr/local/bin/tailwindcss</TailwindCssPath>
+  </PropertyGroup>
+  <Target Name="FindTailwindCss">
+      <Exec Command="where tailwindcss" ConsoleToMSBuild="true" Condition="'$(OS)' == 'Windows_NT'">
+          <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
+      </Exec>
+      <Exec Command="command -v tailwindcss" ConsoleToMSBuild="true" Condition="'$(OS)' != 'Windows_NT'">
+          <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
+      </Exec>
+  </Target>
```
This was the dastardly variable I was trying to load from the environment which locally should be coming from `env.json`. Wouldn't it be best if that's where it was configured and then local and CI would be configured using the same variable?

It's cool that you can search for it and console if it exists but that's also what `build-tailwindcss` does with `which`.

After thinking about it more, while I'd prefer leaning on the bash scritpt, removing it and doing it this way makes the project cross-platform. Which annoys Josh because he hates Windows, but is much better for the client.

Js-Brecht replied:
The only thing I don't like about putting that in `env.json` is that it adds another step for the developer to do to get everything going; especially when there's already a convention that exists for that sort of thing: `$PATH` 

And like you mentioned, the reason I moved it into the MSBuild was to make it cross-platform, though windows does have some mediocre support for bash, and most developers would have that stuff set up (I hope anyway), I thought that this was a more certain way.

I could see clutter in the main build causing confusion being a concern, too, in which case I could move the tailwind stuff out into its own little module (`.targets` file).

Of course, it's all your call.  If you prefer I move to `env.json`, we can definitely go that route.

---

I was wondering what you had been talking about, about your variables not being picked up in the MSBuild.

In order to make that work, I would need to add another section that imports all of the variables into the msbuild, probably as properties.  So you'd wind up that same property, just not visible because it's being pulled in dynamically from a different file.  The other stuff that pulls in the environment variables is for runtime, not the build.

Alternatively, we could set up a wrapper script that bootstraps the build.  Sets up the environment, then calls msbuild for you.

JoshElias replied:
Regarding PATH, I hadn't thought of that. You do usually simply list your dependencies and it's up to the user to make sure they're available from your PATH.

It is also lumped in with a bunch of runtime configuration variables making the build a little more convoluted.

So I agree and no need to load env vars as props to msbuild either.

JoshElias commented:
Yuuuus code review! Thanks Jeremy.

I just read the `BuildTasks` stuff you put in `daemon-dotnet`. Very cool and would be useful for future projects, which you mentioned the other day. It's also a good example of the kind of over engineering that's not needed for this project where the build process is so simple. I'm glad you acknowledged that and left it out of this PR.

JoshElias commented on SmartOrder.csproj:
```diff
@@ -74,13 +81,24 @@
-    <Exec 
-      Command="$(ProjectDir)build-tailwind.sh" 
-    />
+    <Exec Command="$(TailwindCssPath) build -i ./Styles/site.css -o ./wwwroot/css/site.css" />
```
If we're moving away from the bash script, I suppose we'd add these `-i -o` paths to a local `<Property Group>`? No need for them to be configurable from the Environment.

Js-Brecht replied:
To a local what?

Js-Brecht replied:
One of the things that would be really great with the custom tasks approach: we could have MSBuild gather the set of files together using the `<ItemGroup>`, and pass the entire list to the custom task.

Essentially:
* MSBuild scans the project subdirectories for styles
  * In watch mode, we can make it watch for changes in those particular files, and only reprocess the ones that change
* Give the custom task the entire list of files, as well as an output directory
* The custom task takes charge of looping and feeding them to tailwind
* Tailwind then processes them down into their final state, dumping them into the defined output directory.

The way this is currently set up, there's not a very good way to do that, and definitely not one that's cross-platform.

Js-Brecht replied:
We can hard code the style paths for the `-i` and `-o` parameters into a property group for now; I don't see an issue with that.  Later on, we might want to make it more dynamic, so that it uses the `<ItemGroup>` with its wildcard method to find the needed input files, and use those files to determine the proper outputs.

I can look through the builtin tasks and see if there's anything that lets us loop over a set of arguments, and pass each set to a shell process.  I don't think there is, though.

JoshElias replied:
Hardcoding as property group for now sounds good.

JoshElias commented on SmartOrder.csproj:
```diff
@@ -74,13 +81,24 @@
+  </Target>
+  
+  <!-- run tailwindcss watch process by using `dotnet watch run` (or just `dotnet watch`)-->
+  <Target
+    Name="Watch TailwindCss"
+    BeforeTargets="PreBuildEvent"
+    DependsOnTargets="FindTailwindCss"
+    Condition="'$(DotNetWatchBuild)' == 'true'"
+  >
+    <Exec Command="$(TailwindCssPath) build -i ./Styles/site.css -o ./wwwroot/css/site.css --watch" />
   </Target>
```
I had some success relying on msbuild to watch for changes and during it's rebuild it would build tailwind without `--watch`. Maybe the extra watch process is redundant and we can configure msbuild to rebuild on any files that hold tailwind classes that aren't automatically detected already? Like this for example:

```
<ItemGroup>
    <Watch Include="./Styles/**/*.css" />
</ItemGroup>
```

Js-Brecht replied:
That's true, we can definitely do that.  Especially if the tailwind startup is really fast.  If there's a lot of hang time for a large set of styles, then it might be better to keep it as a long running process, sort of like they do for Typescript

Js-Brecht replied:
We can even limit it to only run when the styles themselves change, by adding the `Inputs` attribute to the tailwind `Target`.  That way, it doesn't always have to run tailwind any time any of the source files changes.

JoshElias replied:
I suppose this will also kill the tailwind watch automatically when the parent watch is killed. Lol and I was trying to do that manually...

Considering that tailwind watch will be better tuned for only rebuilding on style changes, let's do it your way.

JoshElias commented:
BTW are really putting `{` on an new line after a function declaration? Weird C# convention IMO.

Js-Brecht reviewed (commented):
> BTW are really putting `{` on an new line after a function declaration? Weird C# convention IMO. 

Personally, I think it's hideous.  Given the opportunity, I would wipe it out of the collective consciousness of the human race.

JoshElias commented:
> > BTW are really putting `{` on an new line after a function declaration? Weird C# convention IMO. 
> 
> Personally, I think it's hideous.  Given the opportunity, I would wipe it out of the collective consciousness of the human race.

Glad we're on the same page there

Js-Brecht committed:
style(global): :wrench: Added .editorconfig for easy local ide formatting configuration

This will allow the editor to more easily know what each particular file uses for white space/indenting

Js-Brecht committed:
fix(build): corrections listed below

* Added support for processing list of css styles
* Added support for a locally installed tailwindcss cli
* Fixed issue with scoped css modules and the `tailwindcss` path
* Removed the `watch` process for tailwindcss

/cc @JoshElias

Js-Brecht reviewed (commented):
@JoshElias update pushed.  Fixed a couple of things with the scoped css.

The error sucks a little bit when tailwindcss isn't installed (just says that the command failed, not why), but the devcontainer will fix that.

After further reflection and research, it makes more sense to remove the special watch target for tailwind, especially after adding support for watching/processing the list of css files.

The tailwind cli itself doesn't support watching multiple files at one time, and from what I was reading, the recommended method for watching a set of files is using something like `chokidar`, watch each input file, and on changes, run the cli on it.  Just like is going to happen with `MSBuild`. 

So I went back to the method you had started to use.  Should work good I think.   Can always revisit if there starts to be a performance hit during development, though I have a feeling there won't be.

JoshElias commented on .editorconfig:
That's niiiice. Thought that only worked with VS. This will be a very welcome change!

JoshElias commented on SmartOrder.csproj:
```diff
@@ -46,59 +41,94 @@
     <PackageReference Include="SmartOrderApi" Version="1.*-*" />
   </ItemGroup>
 
+  <!--
+    Use local node_modules binaries if they exist.
+
+    If there is another directory, say in the CI/CD environment, that tailwindcss
+    is stored, then we should add another <Exec> to this block with the condition
+    that it only run when `'$(CI)' == 'true'`.  That will ensure that this can find
+    the bin anytime
+   -->
+  <Target Name="UseLocalNodeModules" Condition="Exists('node_modules/.bin')">
+    <Exec Command="export PATH=&quot;$(MSBuildProjectDirectory)/node_modules/.bin:$PATH&quot;"  />
+  </Target>
```
Why do we need Node? Is it only to find tailwind if it's installed locally with node? Adding Node this project would add so many configs and other deps...And Javascript. Was so simple with the tailwind executable.

JoshElias replied:
We're supporting Node AND Windows? I'm starting to wonder if you were the right guy for this job... :P 
When I say I use Linux for cross-platform development, what I really mean is I like Linux development.

Js-Brecht replied:
If we're using tailwind, we're using nodejs.  Not really a great way around that unless you want to bundle it up into its own binary, which means it would have to be rebundled for other operating systems and architectures.

Js-Brecht replied:
And yeah, that's for the cases where it gets installed locally.  Standard procedure would be to have it installed locally in the project, so that the project is more portable, but with the container, it wont' be an issue anyway.  It will be doing it's `npm install` dealio automatically, and everything will be really nice and clean.  If we do ever use any other nodejs utilties, then node modules will be needed

JoshElias commented on SmartOrder.csproj:
```diff
@@ -46,59 +41,94 @@
+
+  <Target Name="FindTailwindCss" DependsOnTargets="UseLocalNodeModules">
       <Exec Command="where tailwindcss" ConsoleToMSBuild="true" Condition="'$(OS)' == 'Windows_NT'">
-          <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
+        <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
       </Exec>
       <Exec Command="command -v tailwindcss" ConsoleToMSBuild="true" Condition="'$(OS)' != 'Windows_NT'">
-          <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
+        <Output TaskParameter="ConsoleOutput" PropertyName="TailwindCssPath" />
       </Exec>
   </Target>
```
What a kind developer you are...Letting the user know exactly why they're dumb

JoshElias reviewed (approved):
Looks good to me overall. You'll need to briefly describe how the `IItemList[]` and some other context is pass between Targets but the general idea makes sense.

Would also be nice to know where you learnt it. Was it just the MSBUILD docs? That's a hard subject to firmly grasp.

JoshElias commented:
You're going to have to star the repo before I accept this...

Js-Brecht commented:
> Looks good to me overall. You'll need to briefly describe how the `IItemList[]` and some other context is pass between Targets but the general idea makes sense.
> 
> 
> 
> Would also be nice to know where you learnt it. Was it just the MSBUILD docs? That's a hard subject to firmly grasp.

The power of AI 😁
